src/markov.py

```diff
@@ -24,11 +25,33 @@ def __init__(self,
         self.trans = trans
 
 
+
 def likelihood(x: list[int], mm: MarkovModel) -> float:
     """
     Compute the likelihood of mm given x.
 
     This is the same as the probability of x given mm,
     i.e., P(x ; mm).
     """
-    ...  # FIXME: implement this
+    for i in range(len(x)): 
+        if i == 0: 
+            prob = mm.init_probs[x[i]]
+        else: 
+            prob = prob * mm.trans[x[i-1]][x[i]]
+    return prob
+
+
+def likelihood_log(x: list[int], mm: MarkovModel) -> float:
+    """
+    Compute the likelihood of mm given x.
+
+    This is the same as the probability of x given mm,
+    i.e., P(x ; mm).
+    """
+    for i in range(len(x)): 
+        if i == 0: 
+            prob = math.log(mm.init_probs[x[i]])
+        else: 
+            prob = prob * math.log(mm.trans[x[i-1]][x[i]])
```

Comment:
If you multiply in log-space you are doing exponentiation. $k\log n = \log(n^k)$. You have a multiplication, $nm$, that you want to move to log space, and then you want addition: $\log(nm) = \log n + \log m$.